Add RFC for creation and use of NUMA-constrained arenas (#1559)
uxlfoundation/oneTBB: rfcs/proposed/simplified_numa_support/numa-arenas-creation-and-use.org

* Introduction
Let's consider the example from "Setting the preferred NUMA node" section of the
[[https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Guiding_Task_Scheduler_Execution.html][Guiding Task Scheduler Execution]] page of oneTBB Developer Guide.

** Motivating example
#+begin_src C++
std::vector<tbb::numa_node_id> numa_indexes = tbb::info::numa_nodes();   // [0]
std::vector<tbb::task_arena> arenas(numa_indexes.size());                // [1]
std::vector<tbb::task_group> task_groups(numa_indexes.size());           // [2]

for(unsigned j = 0; j < numa_indexes.size(); j++) {
    arenas[j].initialize(tbb::task_arena::constraints(numa_indexes[j])); // [3]
    arenas[j].execute([&task_groups, &j](){                              // [4]
        task_groups[j].run([](){/*some parallel stuff*/});
    });
}

for(unsigned j = 0; j < numa_indexes.size(); j++) {
    arenas[j].execute([&task_groups, &j](){ task_groups[j].wait(); });   // [5]
}
#+end_src

Usually the users of oneTBB employ this technique to tie oneTBB worker threads
up within NUMA nodes and yet have all the parallelism of a platform utilized.

> The pattern allows to find out how many NUMA nodes are on the system. With that

Reviewer: ```suggestion
The pattern starts by finding the number of NUMA nodes on the system. With that
```

Author: Applied.

> number user creates that many ~tbb::task_arena~ objects, constraining each to a

Reviewer: ```suggestion
number, user creates that many ~tbb::task_arena~ objects, constraining each to a
```

Author: Applied.

  ~task_groups~.
- [3] - Actual tying of ~tbb::task_arena~ instances with corresponding NUMA
  nodes. Note that user needs to make sure the indices of ~tbb::task_arena~
  objects match corresponding indices of NUMA nodes.
- [4] - Actual work submission point. It is relatively easy to make a mistake
  here by using the ~tbb::task_arena::enqueue~ method instead. In this case not
  only the work submission might be done after the synchronization point [5],
  but also the loop counter ~j~ can be mistakenly captured by reference, which
  at least results in submission of the work into incorrect ~tbb::task_group~,
  and at most a segmentation fault, since the loop counter might not exist by
  the time the functor starts its execution.
- [5] - Synchronization point, where user needs to again make sure corresponding
  indices are used. Otherwise, the waiting might be done in unrelated
  ~tbb::task_arena~. It is also possible to mistakenly use
  ~tbb::task_arena::enqueue~ method with the same consequences as were outlined
  in the previous bullet, but since it is a synchronization point, usually the
  blocking call is used.

> The proposal below addresses these issues.

Reviewer 2: I'd prefer it not to address all these issues :)

Specifically, I believe that [4] and [5] are really orthogonal to NUMA, and need to be resolved for `task_arena` in general rather than only for its subclass. And even for the subclass I am not sure that "hiding" a task_group inside is the right thing to do.

Keeping task_group and task_arena separate and merely simplifying the combined use of those with some "syntax sugar" allows creating independent "work queues" in the same arena. Also we can think if a similar approach with a flow graph instead of task_group might be useful. Of course all that can be implemented with the hidden task_group, but likely at the expense of some overhead.

Reviewer 2: The code would be in between of the "verbose" and "concise" variants showed in the document, like this:
```cpp
std::vector<tbb::task_arena> numa_arenas =
    tbb::initialize_constrained_arenas(/*possibly some arguments*/);
std::vector<tbb::task_group> task_groups(numa_arenas.size());

for(unsigned j = 0; j < numa_arenas.size(); j++) {
    numa_arenas[j].enqueue( (){/*some parallel stuff*/}, task_groups[j] );
}

for(unsigned j = 0; j < numa_arenas.size(); j++) {
    numa_arenas[j].wait( task_groups[j] );
}
```
or, with modern C++, like this:
```cpp
std::vector<tbb::task_arena> numa_arenas =
    tbb::initialize_constrained_arenas(/*possibly some arguments*/);
std::vector<tbb::task_group> task_groups(numa_arenas.size());

for(auto& [arena, tg]: std::views::zip(numa_arenas, task_groups)) {
    arena].enqueue( (){/*some parallel stuff*/}, tg );
}

for(auto& [arena, tg]: std::views::zip(numa_arenas, task_groups)) {
    arena.wait( tg );
}
```

Author: I agree that coupling of separate entities is rather a bad thing. Here we would like to improve usability of current interfaces without sacrificing their flexibility. Then the proposal boils down to:
- Introduce the interface that would simplify creation of arenas, each bind to its own NUMA node.
- Introduce the interface that would allow to avoid common mistakes related to loading with work of such arenas.

Reviewer 2: No objections to these two goals.

But I believe the second goal can be achieved without introducing a new class, by extending the existing methods of `task_arena` to better integrate with `task_group`s. That would potentially be useful beyond NUMA scenarios and would give users more control with rather small increase in code complexity.

At the very least, it's an alternative to mention.

Author: Decided to split onto two RFCs. One will simplify constrained arenas instantiation. The other one will deal with submission of work into constrained arenas.

* Proposal
Introduce simplified interface to:
- Contstrain a task arena to specific NUMA node,
- Submit work into constrained task arenas, and
- To wait for completion of the submitted work.

Since the new interface represents a constrained ~tbb::task_arena~ , the
proposed name is ~tbb::constrained_task_arena~. Not including the word "numa"
into the name would allow it for extension in the future for other types of
constraints.

** Usage Example
#+begin_src C++
std::vector<tbb::constrained_task_arena> numa_arenas =
    tbb::initialize_numa_constrained_arenas();

for(unsigned j = 0; j < numa_arenas.size(); j++) {
    numa_arenas[j].enqueue( (){/*some parallel stuff*/} );
}

for(unsigned j = 0; j < numa_arenas.size(); j++) {
    numa_arenas[j].wait();
}
#+end_src

** New arena interface
The example above requires new class named ~tbb::constrained_task_arena~. On one
hand, it is a ~tbb::task_arena~ class that isolates the work execution from
other parallel stuff executed by oneTBB. On the other hand, it is a constrained

> arena that represents an arena associated to a certain NUMA node and allows

Reviewer: Will there be additional functions added to create arenas constrained to core type too? Or will this be exclusively for NUMA?

Author: I guess we need to decide on this whether it is needed right away or can be added later in future RFCs. Initially, I wanted writing RFC that addresses the specific concern about verbose and error-prone API for creation of NUMA-constrained arenas.
